arangod/Auth/UserManager.cpp

```diff
@@ -488,19 +488,38 @@ Result auth::UserManager::enumerateUsers(std::function<bool(auth::User&)>&& func
       }
     }
   }
+  
+  bool triggerUpdate = !toUpdate.empty();
+  
   Result res;
-  {
-    WRITE_LOCKER(writeGuard, _userCacheLock);
-    for (auth::User const& u : toUpdate) {
-      res = storeUserInternal(u, true);
-      if (res.fail()) {
+  do {
+    auto it = toUpdate.begin();
+    while(it != toUpdate.end()) {
+      WRITE_LOCKER(writeGuard, _userCacheLock);
+      res = storeUserInternal(*it, /*replace*/true);
+      
+      if (res.is(TRI_ERROR_ARANGO_CONFLICT) && retryOnConflict) {
+        res.reset();
+        writeGuard.unlock();
+        loadFromDB(); // should be noop iff nothing changed
+        writeGuard.lock();
+        UserMap::iterator it2 = _userCache.find(it->username());
+        if (it2 != _userCache.end()) {
+          auth::User user = it2->second;  // copy user object
+          func(user);
+          *it = std::move(user);
+          continue;
+        }
+      } else if (res.fail()) {
         break;  // do not return, still need to invalidate token cache
       }
+      it = toUpdate.erase(it);
```

Comment:
Will this loop typically process lots of users?
If yes, this may be slow because we are always erasing from the front of the vector, which has linear complexity.
We could iterate the vector in reverse order here if we expect a lot of users in it. But if the vector is typically very small, this seems overkill.